## 1. Import packages

In [1]:
import pandas as pd


## 2. Import dataset

In [10]:
# import dataset from github
df = pd.read_csv('https://raw.githubusercontent.com/ifood/ifood-data-business-analyst-test/master/ml_project1_data.csv', sep=',')

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
1822   8334        1971      Master         Single  33316.0        1   
1194   2245        1969      Master        Married  43641.0        1   
2155   8933        1983    2n Cycle        Married  32146.0        1   
2198   2632        1954  Graduation        Married  50501.0        1   
1521   8643        1971  Graduation       Together  69930.0        0   
2156   8726        1964    2n Cycle       Divorced  41713.0        1   
805    4599        1976      Master        Married  49681.0        0   
1518  10380        1972      Master        Married  37787.0        1   
400    2488        1956    2n Cycle       Divorced  31395.0        1   
1831   9860        1959  Graduation       Together  44911.0        0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
1822         1  2013-10-04       34        79  ...                  6   
1194         1  2013-04-22       50        57  ...                  6   
2155         0  2014-05-12       16        26  ...                  6   
2198         1  2013-06-18       10       297  ...                  6   
1521         0  2013-05-26       21       252  ...                  3   
2156         1  2014-04-29       57        77  ...                  6   
805          2  2013-11-04       66       411  ...                  7   
1518         0  2013-09-20       50        40  ...                  8   
400          1  2013-03-23       80        23  ...                  6   
1831         1  2013-03-15       11       159  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
1822             0             0             0             0             0   
1194             0             0             0             0             0   
2155             0             0             0             0             0   
2198             1             0             0             0             0   
1521             0             0             0             0             0   
2156             0             0             0             0             0   
805              0             0             0             0             0   
1518             0             0             0             0             0   
400              0             0             0             0             0   
1831             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
1822         0              3         11         0  
1194         0              3         11         0  
2155         0              3         11         0  
2198         0              3         11         1  
1521         0              3         11         0  
2156         0              3         11         0  
805          0              3         11         0  
1518         0              3         11         0  
400          0              3         11         0  
1831         0              3         11         0  

[10 rows x 29 columns]

## 3. Exploratory Analysis
Notes:
1. shape = (2240,29)
2. Income has NaN values
3.


### 3.1 Let's see quantity of samples that has Income = NaN (None)

In [7]:
df[df['Income'].isnull()]

ID  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
10     1994        1983  Graduation        Married     NaN        1         0   
27     5255        1986  Graduation         Single     NaN        1         0   
43     7281        1959         PhD         Single     NaN        0         0   
48     7244        1951  Graduation         Single     NaN        2         1   
58     8557        1982  Graduation         Single     NaN        1         0   
71    10629        1973    2n Cycle        Married     NaN        1         0   
90     8996        1957         PhD        Married     NaN        2         1   
91     9235        1957  Graduation         Single     NaN        1         1   
92     5798        1973      Master       Together     NaN        0         0   
128    8268        1961         PhD        Married     NaN        0         1   
133    1295        1963  Graduation        Married     NaN        0         1   
312    2437        1989  Graduation        Married     NaN        0         0   
319    2863        1970  Graduation         Single     NaN        1         2   
1379  10475        1970      Master       Together     NaN        0         1   
1382   2902        1958  Graduation       Together     NaN        1         1   
1383   4345        1964    2n Cycle         Single     NaN        1         1   
1386   3769        1972         PhD       Together     NaN        1         0   
2059   7187        1969      Master       Together     NaN        1         1   
2061   1612        1981         PhD         Single     NaN        1         0   
2078   5079        1971  Graduation        Married     NaN        1         1   
2079  10339        1954      Master       Together     NaN        0         1   
2081   3117        1955  Graduation         Single     NaN        0         1   
2084   5250        1943      Master          Widow     NaN        0         0   
2228   8720        1978    2n Cycle       Together     NaN        0         0   

     Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
10    2013-11-15       11         5  ...                  7             0   
27    2013-02-20       19         5  ...                  1             0   
43    2013-11-05       80        81  ...                  2             0   
48    2014-01-01       96        48  ...                  6             0   
58    2013-06-17       57        11  ...                  6             0   
71    2012-09-14       25        25  ...                  8             0   
90    2012-11-19        4       230  ...                  9             0   
91    2014-05-27       45         7  ...                  7             0   
92    2013-11-23       87       445  ...                  1             0   
128   2013-07-11       23       352  ...                  6             0   
133   2013-08-11       96       231  ...                  4             0   
312   2013-06-03       69       861  ...                  3             0   
319   2013-08-23       67       738  ...                  7             0   
1379  2013-04-01       39       187  ...                  5             0   
1382  2012-09-03       87        19  ...                  5             0   
1383  2014-01-12       49         5  ...                  7             0   
1386  2014-03-02       17        25  ...                  7             0   
2059  2013-05-18       52       375  ...                  3             0   
2061  2013-05-31       82        23  ...                  6             0   
2078  2013-03-03       82        71  ...                  8             0   
2079  2013-06-23       83       161  ...                  6             0   
2081  2013-10-18       95       264  ...                  7             0   
2084  2013-10-30       75       532  ...                  1             0   
2228  2012-08-12       53        32  ...                  0             0   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedC

Notes:


### 3.2. Analyzing 10 samples (5 with target = 1 and 5 with target = 0)
Objective: find nice features and bad features. Maybe remove features to use better data

In [31]:
df_3_2 = df.copy()
df_3_2['Age'] = df.apply(lambda x: 2014 - x['Year_Birth'], axis=1)
samples = df_3_2.sort_values(by=['Response'], ascending=False).groupby('Response').head(5)
samples

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1084   6072        1970      Master         Single  75345.0        0   
1000  10473        1956         PhD          Widow  78028.0        0   
1001   7962        1987         PhD         Single  95169.0        0   
1002   9984        1981    2n Cycle        Married  56337.0        1   
106    7533        1964  Graduation        Married  49096.0        1   
105    8373        1979       Basic       Together  24594.0        1   
1587   2471        1965  Graduation        Married  61482.0        0   
1718  10640        1974  Graduation       Together  44392.0        1   
1588   5892        1980  Graduation        Married  34968.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  AcceptedCmp3  \
0            0  2012-09-04       58       635  ...             0   
1084         0  2013-02-02       16       918  ...             0   
1000         1  2012-09-10       38       158  ...             0   
1001         0  2013-10-09        1      1285  ...             0   
1002         1  2013-03-27       25       349  ...             1   
106          1  2013-09-24       15       144  ...             0   
105          0  2013-12-10       94         1  ...             0   
1587         0  2014-06-27       39        48  ...             0   
1718         1  2013-09-05       71        87  ...             0   
1588         0  2013-04-10       11       158  ...             0   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0                0             0             0             0         0   
1084             0             0             1             0         0   
1000             1             1             0             0         0   
1001             0             1             1             0         0   
1002             0             0             1             0         0   
106              0             0             0             0         0   
105              0             0             0             0         0   
1587             0             0             0             0         0   
1718             0             0             0             0         0   
1588             0             0             0             0         0   

      Z_CostContact  Z_Revenue  Response  Age  
0                 3         11         1   57  
1084              3         11         1   44  
1000              3         11         1   58  
1001              3         11         1   27  
1002              3         11         1   33  
106               3         11         0   50  
105               3         11         0   35  
1587              3         11         0   49  
1718              3         11         0   40  
1588              3         11         0   34  

[10 rows x 30 columns]

Notes (Initial hypotheses):
- Columns correlated to Response: MntFruits,Meat,Fish,Hold; NumWebPurchases,CatalogPurchases; Education; Marital_Status
- Only 2 negative customers accepted the Campaign2
- None insight was found with positive customers related to older Campaigns

### 3.3 Lets see the statistics about data

In [8]:
df.describe()

ID   Year_Birth         Income      Kidhome     Teenhome  \
count   2240.000000  2240.000000    2216.000000  2240.000000  2240.000000   
mean    5592.159821  1968.805804   52247.251354     0.444196     0.506250   
std     3246.662198    11.984069   25173.076661     0.538398     0.544538   
min        0.000000  1893.000000    1730.000000     0.000000     0.000000   
25%     2828.250000  1959.000000   35303.000000     0.000000     0.000000   
50%     5458.500000  1970.000000   51381.500000     0.000000     0.000000   
75%     8427.750000  1977.000000   68522.000000     1.000000     1.000000   
max    11191.000000  1996.000000  666666.000000     2.000000     2.000000   

           Recency     MntWines    MntFruits  MntMeatProducts  \
count  2240.000000  2240.000000  2240.000000      2240.000000   
mean     49.109375   303.935714    26.302232       166.950000   
std      28.962453   336.597393    39.773434       225.715373   
min       0.000000     0.000000     0.000000         0.000000   
25%      24.000000    23.750000     1.000000        16.000000   
50%      49.000000   173.500000     8.000000        67.000000   
75%      74.000000   504.250000    33.000000       232.000000   
max      99.000000  1493.000000   199.000000      1725.000000   

       MntFishProducts  ...  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
count      2240.000000  ...        2240.000000   2240.000000   2240.000000   
mean         37.525446  ...           5.316518      0.072768      0.074554   
std          54.628979  ...           2.426645      0.259813      0.262728   
min           0.000000  ...           0.000000      0.000000      0.000000   
25%           3.000000  ...           3.000000      0.000000      0.000000   
50%          12.000000  ...           6.000000      0.000000      0.000000   
75%          50.000000  ...           7.000000      0.000000      0.000000   
max         259.000000  ...          20.000000      1.000000      1.000000   

       AcceptedCmp5  AcceptedCmp1  AcceptedCmp2     Complain  Z_CostContact  \
count   2240.000000   2240.000000   2240.000000  2240.000000         2240.0   
mean       0.072768      0.064286      0.013393     0.009375            3.0   
std        0.259813      0.245316      0.114976     0.096391            0.0   
min        0.000000      0.000000      0.000000     0.000000            3.0   
25%        0.000000      0.000000      0.000000     0.000000            3.0   
50%        0.000000      0.000000      0.000000     0.000000            3.0   
75%        0.000000      0.000000      0.000000     0.000000            3.0   
max        1.000000      1.000000      1.000000     1.000000            3.0   

       Z_Revenue     Response  
count     2240.0  2240.000000  
mean        11.0     0.149107  
std          0.0     0.356274  
min         11.0     0.000000  
25%         11.0     0.000000  
50%         11.0     0.000000  
75%         11.0     0.000000  
max         11.0     1.000000  

[8 rows x 26 columns]

Notes:


### 3.4 Lets see the balancement of the dataset

In [11]:
df['Response'].value_counts()

0    1906
1     334
Name: Response, dtype: int64

Notes:
- The dataset is imbalanced
- Percentage of customers that responded to campaign: 15%
- Percentage of customers that not responded to campaign: 85%
- Therefore, the future model will be good to predict a customer that will not respond to a campaign and be bad to predict if customer will respond to campaign

### 3.4 Lets see the mean of each column based on label

In [32]:
df_3_4 = df.groupby('Response').mean()
df_3_4


ID   Year_Birth        Income   Kidhome  Teenhome  \
Response                                                               
0         5622.009444  1968.698846  50839.132767  0.462225  0.541448   
1         5421.820359  1969.416168  60209.675676  0.341317  0.305389   

            Recency    MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
Response                                                                       
0         51.514690  269.104407  24.216684       144.624344        34.980063   
1         35.383234  502.703593  38.203593       294.353293        52.050898   

          ...  NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  \
Response  ...                                                       
0         ...           5.736621           5.320567      0.045121   
1         ...           6.095808           5.293413      0.230539   

          AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
Response                                                                     
0             0.055089      0.037251      0.034103      0.005247  0.009444   
1             0.185629      0.275449      0.236527      0.059880  0.008982   

          Z_CostContact  Z_Revenue  
Response                            
0                   3.0       11.0  
1                   3.0       11.0  

[2 rows x 25 columns]

Notes:
- Year_Birth, **Income**, **Mnt.***,  Num.* - NumWebVisitsMonth, **AcceptedCmp[1-5]** was higher for **Response = 1**
- **Kidhome**, **Teenhome**, **Recency**, NumWebVisitsMonth, Complain was higher for **Response = 0**
- OBS: for the Num.* columns, NumWebPurchases, NumCatalogPurchases, NumStorePurchases have the most significant diff (>0.15)
